pkg/hub/submarinerbroker/brokerCRDsController.go

```diff
@@ -0,0 +1,128 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+)
+
+const (
+	configCRDName = "submarinerconfigs.submarineraddon.open-cluster-management.io"
+)
+
+var (
+	staticCRDFiles = []string{
+		"manifests/broker/submariner.io_clusters_crd.yaml",
+		"manifests/broker/submariner.io_endpoints_crd.yaml",
+		"manifests/broker/submariner.io_gateways_crd.yaml",
+		"manifests/broker/submariner.io_lighthouse.serviceimports_crd.yaml",
+		"manifests/broker/x-k8s.io_multicluster.serviceimports_crd.yaml",
+	}
+
+	brokerCRDNames = []string{
+		"clusters.submariner.io",
+		"endpoints.submariner.io",
+		"gateways.submariner.io",
+		"serviceimports.lighthouse.submariner.io",
+		"serviceimports.multicluster.x-k8s.io",
+	}
+)
+
+type brokerCRDsConfig struct {
+	ConfigCRDUID types.UID
+}
+
+type submarinerBrokerCRDsController struct {
+	crdClient     apiextensionsclientset.Interface
+	eventRecorder events.Recorder
+}
+
+func NewSubmarinerBrokerCRDsController(
+	crdClient apiextensionsclientset.Interface,
+	crdInformer apiextensionsinformers.CustomResourceDefinitionInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &submarinerBrokerCRDsController{
+		crdClient:     crdClient,
+		eventRecorder: recorder.WithComponentSuffix("submariner-broker-crds-controller"),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, crdInformer.Informer()).
+		WithSync(c.sync).
+		ToController("SubmarinerBrokerCRDsController", recorder)
+}
+
+func (c *submarinerBrokerCRDsController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	crdName := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling ConfigCRD %q", crdName)
+
+	if crdName != configCRDName {
+		return nil
+	}
+
+	configCRD, err := c.crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(ctx, crdName, v1.GetOptions{})
+	if errors.IsNotFound(err) {
+		return c.cleanUp(ctx)
+	}
+	if err != nil {
+		return err
+	}
+
+	crdsConfig := brokerCRDsConfig{
+		ConfigCRDUID: configCRD.GetUID(),
+	}
+
+	// Apply CRDs
+	clientHolder := helpers.NewCRDClientHolder().WithAPIExtensionsClient(c.crdClient)
+	applyResults := helpers.ApplyCRDDirectly(
+		clientHolder,
+		syncCtx.Recorder(),
+		func(name string) ([]byte, error) {
+			return assets.MustCreateAssetFromTemplate(name, bindata.MustAsset(filepath.Join("", name)), crdsConfig).Data, nil
+		},
+		staticCRDFiles...,
+	)
+
+	errs := []error{}
+	for _, result := range applyResults {
+		if result.Error != nil {
+			errs = append(errs, fmt.Errorf("%q (%T): %v", result.File, result.Type, result.Error))
+		}
+	}
+
+	return operatorhelpers.NewMultiLineAggregate(errs)
+}
+
+func (c *submarinerBrokerCRDsController) cleanUp(ctx context.Context) error {
+	errs := []error{}
+	for _, crd := range brokerCRDNames {
+		err := c.crdClient.ApiextensionsV1().CustomResourceDefinitions().Delete(ctx, crd, metav1.DeleteOptions{})
```

Comment:
do we get into this code? the crd is remove only when operator is deleted from olm.  Will olm guarantee the sequence that crd is delete at first, then operator?

Reply:
not sure...

Reply:
I think CRD is the last one to be deleted in uninstall Operator process (refer https://github.com/operator-framework/olm-book/blob/master/docs/uninstall-an-operator.md)  
cleanup is not needed here.